circuits/cpp/barretenberg/cpp/src/barretenberg/honk/proof_system/permutation_library.hpp

```diff
@@ -0,0 +1,165 @@
+#pragma once
+#include "barretenberg/honk/sumcheck/sumcheck.hpp"
+#include "barretenberg/plonk/proof_system/proving_key/proving_key.hpp"
+#include "barretenberg/polynomials/polynomial.hpp"
+#include <typeinfo>
+
+namespace proof_system::honk::permutation_library {
+
+/**
+ * @brief Compute a permutation grand product polynomial Z_perm(X)
+ * *
+ * @details
+ * Z_perm may be defined in terms of its values  on X_i = 0,1,...,n-1 as Z_perm[0] = 1 and for i = 1:n-1
+ *                  relation::numerator(j)
+ * Z_perm[i] = ∏ --------------------------------------------------------------------------------
+ *                  relation::denominator(j)
+ *
+ * where ∏ := ∏_{j=0:i-1}
+ *
+ * The specific algebraic relation used by Z_perm is defined by Flavor::GrandProductRelations
+ *
+ * For example, in Flavor::Standard the relation describes:
+ *
+ *                  (w_1(j) + β⋅id_1(j) + γ) ⋅ (w_2(j) + β⋅id_2(j) + γ) ⋅ (w_3(j) + β⋅id_3(j) + γ)
+ * Z_perm[i] = ∏ --------------------------------------------------------------------------------
+ *                  (w_1(j) + β⋅σ_1(j) + γ) ⋅ (w_2(j) + β⋅σ_2(j) + γ) ⋅ (w_3(j) + β⋅σ_3(j) + γ)
+ * where ∏ := ∏_{j=0:i-1} and id_i(X) = id(X) + n*(i-1)
+ *
+ * For Flavor::Ultra both the UltraPermutation and Lookup grand products are computed by this method.
+ *
+ * The grand product is constructed over the course of three steps.
+ *
+ * For expositional simplicity, write Z_perm[i] as
+ *
+ *                A(j)
+ * Z_perm[i] = ∏ --------------------------
+ *                B(h)
+ *
+ * Step 1) Compute 2 length-n polynomials A, B
+ * Step 2) Compute 2 length-n polynomials numerator = ∏ A(j), nenominator = ∏ B(j)
+ * Step 3) Compute Z_perm[i + 1] = numerator[i] / denominator[i] (recall: Z_perm[0] = 1)
+ *
+ * Note: Step (3) utilizes Montgomery batch inversion to replace n-many inversions with
+ */
+template <typename Flavor, typename PermutationRelation>
+void compute_permutation_grand_product(const size_t circuit_size,
+                                       auto& full_polynomials,
+                                       proof_system::RelationParameters<typename Flavor::FF>& relation_parameters)
+{
+    using FF = typename Flavor::FF;
+    using Polynomial = typename Flavor::Polynomial;
+    using ValueAccumulatorsAndViews = typename PermutationRelation::ValueAccumulatorsAndViews;
+
+    // Allocate numerator/denominator polynomials that will serve as scratch space
+    // TODO(zac) we can re-use the permutation polynomial as the numerator polynomial. Reduces readability
+    Polynomial numerator = Polynomial{ circuit_size };
+    Polynomial denominator = Polynomial{ circuit_size };
+
+    // Step (1)
+    // Populate `numerator` and `denominator` with the algebra described by PermutationRelation
+    const size_t num_threads = circuit_size >= get_num_cpus_pow2() ? get_num_cpus_pow2() : 1;
```

Comment:
If I have 2 cores, parallelism will start at circuit size 2. This will probably not be very efficient.
Also, if you are only using powers of 2, while it is more convenient, on a laptop with, for example, 14 cores, this will not use almost half

Reply:
re: the powers of 2 thing. We do this across the entire codebase with our current parralelism. Agreed that it is not efficient, I will make a TODO.

Will also update to make sure there is a minimum circuit size before we parralelize